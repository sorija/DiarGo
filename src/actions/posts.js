import database from '../firebase/firebase';
import moment from 'moment';

export const addPost = (post) => ({
  type: 'ADD_POST',
  post
});

export const startAddPost = (postData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      content = '',
      createdAt = 0
    } = postData;
    // save data
    const post = { title, content, createdAt };
    return database.ref(`users/${uid}/posts`).push(post).then((ref) => {
      dispatch(addPost({
        id: ref.key,
        ...post
      }));
    });
  };
};

export const setPosts = (posts) => ({
  type: 'SET_POSTS',
  posts
});

export const startSetPosts = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts`).once('value').then((snapshot) => {
      const posts = [];
      snapshot.forEach((childSnapshot) => {
        posts.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setPosts(posts));
    });
  };
};

export const removePost = ({ id } = {}) => ({
  type: 'REMOVE_POST',
  id
});

export const startRemovePost = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts/${id}`).remove().then(() => {
      dispatch(removePost({ id }));
    });
  };
};

export const editPost = (id, updates) => ({
  type: 'EDIT_POST',
  id,
  updates
});

export const startEditPost = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts/${id}`).update(updates).then(() => {
      dispatch(editPost(id, updates));
    });
  };
};
