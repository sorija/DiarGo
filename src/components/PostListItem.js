import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const PostListItem = ({ id, title, content, createdAt }) => (
  <Link className="list-item" to={`/read/${id}`}>
    <span className="list-item__date">{moment(createdAt).format('MMMM Do, YYYY [at] h:mm:ss A')}</span>
    <h2 className="list-item__title">{title}</h2>
  </Link>
);

export default PostListItem;
