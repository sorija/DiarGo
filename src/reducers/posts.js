const postsReducer = (state = [], action) => {
  switch (action.type) {
    case 'ADD_POST':
      return [...state, action.post];
    case 'SET_POSTS':
      return action.posts;
    case 'EDIT_POST':
    return state.map((post) => {
      if (post.id === action.id) {
        return {
          ...post,
          ...action.updates
        };
      } else {
        return post;
      };
    });
    case 'REMOVE_POST':
      return state.filter(({ id }) => id !== action.id);
    default:
      return state;
  }
};

export default postsReducer;
