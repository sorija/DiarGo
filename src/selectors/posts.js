import moment from 'moment';

const selectPosts = (posts, { text, date }) => {
  return posts.filter((post) => {
    const createdAtMoment = moment(post.createdAt);
    const textMatch = post.title.toLowerCase().includes(text.toLowerCase());
    const dateMatch = date ? date.isSame(createdAtMoment, 'day') : true;
    
    return textMatch && dateMatch;
  });
};

export default selectPosts;
