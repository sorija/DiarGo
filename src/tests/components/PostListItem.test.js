import React from 'react';
import { shallow } from 'enzyme';
import PostListItem from '../../components/PostListItem';
import posts from '../fixtures/posts';

test('should render PostListItem with a post', () => {
  const wrapper = shallow(<PostListItem {...posts[0]}/>);
  expect(wrapper).toMatchSnapshot();
});
