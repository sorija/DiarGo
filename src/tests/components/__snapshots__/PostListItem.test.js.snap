// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PostListItem with a post 1`] = `
<Link
  replace={false}
  to="/edit/1"
>
  <h2>
    Test Post Title
  </h2>
  <p>
    Test Post Content
  </p>
</Link>
`;
